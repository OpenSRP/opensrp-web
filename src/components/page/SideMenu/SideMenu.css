.side-nav-item {
  color: #3A484F;
  font-family: Roboto;
  font-weight: 500;
  line-height: 20px;
  text-align: left;
}

.side-nav-active {
  color: #0192D4 !important;
  font-family: Roboto;
  font-weight: 500;
  text-align: left;
}