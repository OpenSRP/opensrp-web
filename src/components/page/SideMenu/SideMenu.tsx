import * as React from 'react';

class SideMenu extends React.Component<{}, {}> {
  constructor(props: {}) {
    super(props);
  }

  public render() {
    return <div>Side Menu</div>;
  }
}

export default SideMenu;
