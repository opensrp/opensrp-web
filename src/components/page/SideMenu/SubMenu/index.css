.side-collapse-nav.nav {
  cursor: pointer;
  padding: 4px 12px 2px 20px;
  max-width: initial !important;
}

.nav-item-extend {
  width: 100%;
  padding-left: 20px;
}

.collapse-menu-title {
  padding-left: 12px;
}

.active-collapse-menu {
  color: #0192d4 !important;
}

.active-collapse-menu:hover {
  color: #0192d4 !important;
}

.side-nav-active {
  border-right: 4px solid #209fd9;
}

.side-menu-container .side-menu-extend ul.nav .collapse.show {
  padding-right: initial;
  padding-left: 33px;
}

.side-nav-item.nav {
  max-width: initial;
  background-color: initial;
  font-family: 'Roboto';
  font-size: 14px;
  font-weight: 500;
}

.side-nav-item.nav:hover {
  color: #0192d4 !important;
}

.side-menu-container .collapse {
  padding-right: initial;
  padding-left: 33px;
}

.active-collapse-menu path {
  fill: #0192d4;
}

path {
  fill: #1b314b;
}
