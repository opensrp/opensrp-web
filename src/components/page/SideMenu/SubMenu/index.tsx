import { IconProp } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import classNames from 'classnames';
import React, { Fragment } from 'react';
import { RouteComponentProps, withRouter } from 'react-router';
import { NavLink } from 'react-router-dom';
import { Collapse, Nav, NavItem } from 'reactstrap';
import './index.css';

/** interface for a module i.e group of related pages
 * @property {string} parentNav - Page navigation for the module, collapses to reveal the page links
 * @property {PageLink[]} childNavs - pageLinks to pages under this navigational module
 */
export interface NavigationModule {
  parentNav: ModulePageLink;
  childNavs: PageLink[];
}

/** interface for objs describing links to different pages
 * @property {string} label - the link's displayable text
 * @property {string} url - URL to get redirected on click
 */
export interface PageLink {
  label: string;
  url: string;
}

/** interface for the module page links, module represents collection of related page links
 * @property {string} label - text label visible on menu item
 * @property {IconProp} icon- fontAwesome icon displayed on menu item
 * @property {string} url - url that module link will redirect to
 */
export interface ModulePageLink {
  label: string;
  icon: IconProp;
  url?: string;
}

/** props for this component
 * @property {ModulePageLink} parentNav - object describing navigation data for the module page link
 * @property {PageLink[]} childNavs - array of the pageNavigation links grouped under this navigation module
 * @property {string} collapsedModuleLabel - string literal with the label value of currently collapsed module navigation link
 * @property {setCollapsedModuleLabel} setCollapsedModuleLabel - callback to parent component;
 *            changes the collapsedModuleLabel entry in the parent component
 */
export interface SubMenuProps {
  parentNav: ModulePageLink;
  childNavs: PageLink[];
  collapsedModuleLabel: string;
  setCollapsedModuleLabel?: SetCollapsedModuleLabel;
  customIcon?: React.FunctionComponent<React.SVGProps<SVGSVGElement>>;
}

/** this components state
 * @property {boolean} collapseMenu - whether the child navigation links should be collapsed or hidden
 */
export interface SubMenuState {
  isActiveModule: boolean;
}

/** takes a string and updates the current collapsed-module-label in the parent component
 * @param {string} label - label of clicked navigation module pagelink
 */
type SetCollapsedModuleLabel = (label: string) => void;

/** Default props */
export const defaultSubMenuProps: SubMenuProps = {
  childNavs: [],
  collapsedModuleLabel: '',
  parentNav: {
    icon: ['far', 'user'],
    label: 'home',
  },
};

/** intersection of all types describing the props */
type subMenuPropsTypes = SubMenuProps & RouteComponentProps;

export class SubMenu extends React.Component<subMenuPropsTypes, SubMenuState> {
  public static defaultProps = defaultSubMenuProps;

  public render() {
    const { childNavs, parentNav, collapsedModuleLabel, location } = this.props;
    const pathName = location.pathname;
    /** whether to collapse the child pages for this navigation module */
    const collapseMenu: boolean = collapsedModuleLabel === parentNav.label;
    /** whether the navigation module is currently active; judging from the history.location pathname
     * set to true if one of the module navigation's children has a url similar to pathname or
     * if the url of the parentNav is similar to the pathname
     */
    const isActiveModule: boolean =
      childNavs.map(nav => nav.url).indexOf(pathName!) > -1 || parentNav.url === pathName;

    const moduleLinkJsx = (
      <Fragment>
        {this.props.customIcon ? (
          <this.props.customIcon />
        ) : (
          <FontAwesomeIcon icon={parentNav.icon} size="lg" />
        )}
        <span className="collapse-menu-title"> {parentNav.label} </span>
      </Fragment>
    );

    const moduleLinkClassName = classNames('nav', 'nav-link', 'side-nav-item', {
      'active-collapse-menu': isActiveModule,
    });

    return (
      <div id={`sub-menu-${parentNav.label.replace(' ', '-')}`}>
        <Nav
          id={parentNav.label.replace(' ', '-')}
          className="side-collapse-nav"
          onClick={this.setModuleLabel}
        >
          <NavItem>
            {parentNav.url ? (
              <NavLink to={parentNav.url} className={moduleLinkClassName}>
                {moduleLinkJsx}
              </NavLink>
            ) : (
              <span className={moduleLinkClassName}>{moduleLinkJsx}</span>
            )}
          </NavItem>
        </Nav>

        <Collapse isOpen={collapseMenu}>
          {childNavs.map((childNavObj, key) => {
            return (
              <Nav key={'child-nav-' + key}>
                <NavItem className="nav-item-extend">
                  <NavLink
                    to={childNavObj.url}
                    className="nav-link side-nav-item"
                    activeClassName="side-nav-active"
                  >
                    <span> {childNavObj.label} </span>
                  </NavLink>
                </NavItem>
              </Nav>
            );
          })}
        </Collapse>
      </div>
    );
  }

  /** updates parent component with the label of the currently collapsed navigation module */
  private setModuleLabel = () => {
    const labelAlreadySet = this.props.collapsedModuleLabel === this.props.parentNav.label;
    labelAlreadySet && this.props.setCollapsedModuleLabel !== undefined
      ? this.props.setCollapsedModuleLabel('')
      : this.props.setCollapsedModuleLabel!(this.props.parentNav.label);
  };
}

export default withRouter(SubMenu);
