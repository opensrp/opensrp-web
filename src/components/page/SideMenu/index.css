.side-nav-item {
  color: #3a484f;
  font-weight: 500;
  line-height: 20px;
  text-align: left;
}

.side-nav-active {
  background: rgba(1, 146, 212, 0.1);
  color: #0192d4 !important;
  font-weight: 500;
  text-align: left;
}

.side-menu-extend {
  margin: 0;
  padding: 16px 0 0 0;
}

.main-container .hidden-container {
  display: none !important;
}

.side-menu-extend ul.nav {
  background: initial;
  background-color: initial;
  max-width: initial;
  width: 100%;
}

.main-container .sidebar {
  width: 300px;
}

.sidebar .side-menu-extend a:hover {
  text-decoration: none;
  color: #0192d4;
}

.sidebar .side-menu-container {
  position: relative;
  width: 100%;
  height: 100%;
  flex: 1 1;
  border: 1px;
  border-right-style: solid;
  border-right-color: #dfe3e8;
  background: #f8fafe;
}

.sidebar .side-menu-container .row {
  margin: auto;
}

.side-menu-container .side-menu-extend ul.side-collapse-nav .collapse {
  padding-right: initial;
  padding-left: 33px;
}
