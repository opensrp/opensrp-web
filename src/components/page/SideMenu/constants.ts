import { ModulePageLink, NavigationModule, PageLink } from './SubMenu';

import {
  ADMIN,
  ALL_CLIENTS,
  ANALYSIS,
  ANALYSIS_URL,
  ANC,
  ANC_URL,
  CHILD,
  CHILD_URL,
  CLIENT_RECORDS,
  CLIENT_URL,
  COMPARTMENTS,
  COMPARTMENTS_URL,
  HOME,
  HOME_URL,
  HOUSEHOLD,
  HOUSEHOLD_URL,
  LOCATIONS,
  LOCATIONS_URL,
  LOGFACE,
  LOGFACE_URL,
  NBC_AND_PNC,
  NBC_AND_PNC_URL,
  NUTRITION,
  NUTRITION_URL,
  PREGNANCY,
  PREGNANCY_URL,
  REPORTS,
  REPORTS_URL,
  ROLE_URL,
  ROLES,
  TEAM_URL,
  TEAMS,
  USER_URL,
  USERS,
} from '../../../constants';

import { IconProp } from '@fortawesome/fontawesome-svg-core';

import {
  ENABLE_ANC_PAGE,
  ENABLE_CHILD_PAGE,
  ENABLE_CLIENT_PAGE,
  ENABLE_HOUSEHOLD_PAGE,
  ENABLE_LOCATION_PAGE,
  ENABLE_ROLE_PAGE,
  ENABLE_TEAM_PAGE,
  ENABLE_USER_PAGE,
} from '../../../configs/env';

// Page links

export const LOG_FACE: PageLink = {
  label: 'logface',
  url: '/logface',
};

export const CompartMents: PageLink = {
  label: 'compartments',
  url: '/compartments',
};

export const Analysis: PageLink = {
  label: 'Analysis',
  url: '/analysis',
};
export const CLIENT_PAGE_NAVIGATION: PageLink = {
  label: ALL_CLIENTS,
  url: CLIENT_URL,
};
export const HOUSEHOLD_PAGE_NAVIGATION: PageLink = {
  label: HOUSEHOLD,
  url: HOUSEHOLD_URL,
};
export const ANC_PAGE_NAVIGATION: PageLink = {
  label: ANC,
  url: ANC_URL,
};
export const CHILD_PAGE_NAVIGATION: PageLink = {
  label: CHILD,
  url: CHILD_URL,
};

export const USERS_PAGE_NAVIGATION: PageLink = {
  label: USERS,
  url: USER_URL,
};
export const ROLES_PAGE_NAVIGATION: PageLink = {
  label: ROLES,
  url: ROLE_URL,
};
export const TEAMS_PAGE_NAVIGATION: PageLink = {
  label: TEAMS,
  url: TEAM_URL,
};
export const LOCATIONS_PAGE_NAVIGATION: PageLink = {
  label: LOCATIONS,
  url: LOCATIONS_URL,
};

export const PREGNANCY_PAGE_NAVIGATION: PageLink = {
  label: PREGNANCY,
  url: PREGNANCY_URL,
};

export const NUTRITION_PAGE_NAVIGATION: PageLink = {
  label: NUTRITION,
  url: NUTRITION_URL,
};

export const NEWBORN_AND_POSTNATAL_PAGE_NAVIGATION: PageLink = {
  label: NBC_AND_PNC,
  url: NBC_AND_PNC_URL,
};

export const COMPARTMENTS_PAGE_NAVIGATION: PageLink = {
  label: COMPARTMENTS,
  url: COMPARTMENTS_URL,
};

export const LOGFACE_PAGE_NAVIGATION: PageLink = {
  label: LOGFACE,
  url: LOGFACE_URL,
};

export const ANALYSIS_PAGE_NAVIGATION: PageLink = {
  label: ANALYSIS,
  url: ANALYSIS_URL,
};

// icons
export const homeNavIcon: IconProp = ['fas', 'home'];
export const pregancyNavIcon: IconProp = ['far', 'user'];
export const nbcandpncNavIcon: IconProp = ['far', 'user'];
export const nutritionNavIcon: IconProp = ['far', 'user'];
export const newbornAndPostnatalNavIcon: IconProp = ['far', 'user'];
export const clientModuleNavIcon: IconProp = ['far', 'user'];
export const reportModuleNavIcon: IconProp = ['fas', 'chart-line'];
export const adminModuleNavIcon: IconProp = ['fas', 'cog'];

// Navigation links for navigation module. collapses when clicked to display the pageLinks

export const HOME_PARENT_NAV: ModulePageLink = {
  icon: homeNavIcon,
  label: HOME,
  url: HOME_URL,
};

export const NBC_AND_PNC_PARENT_NAV: ModulePageLink = {
  icon: nbcandpncNavIcon,
  label: NBC_AND_PNC,
  url: NBC_AND_PNC_URL,
};

export const NUTRITION_PARENT_NAVIGATION: ModulePageLink = {
  icon: nutritionNavIcon,
  label: NUTRITION,
  url: NUTRITION_URL,
};

export const PREGNANCY_MODULE_PARENT_NAV: ModulePageLink = {
  icon: pregancyNavIcon,
  label: PREGNANCY,
  url: PREGNANCY_URL,
};

export const CLIENT_MODULE_PARENT_NAV: ModulePageLink = {
  icon: clientModuleNavIcon,
  label: CLIENT_RECORDS,
};
export const REPORT_MODULE_PARENT_NAV: ModulePageLink = {
  icon: reportModuleNavIcon,
  label: REPORTS,
  url: REPORTS_URL,
};
export const ADMIN_MODULE_PARENT_NAV: ModulePageLink = {
  icon: adminModuleNavIcon,
  label: ADMIN,
};

// nav module objects
export const HOME_NAVIGATION_MODULE: NavigationModule = {
  childNavs: [],
  parentNav: HOME_PARENT_NAV,
};

export const PREGNANCY_NAVIGATION_MODULE: NavigationModule = {
  childNavs: [
    LOGFACE_PAGE_NAVIGATION,
    COMPARTMENTS_PAGE_NAVIGATION,
    ANALYSIS_PAGE_NAVIGATION,
  ].filter((childNav): childNav is PageLink => typeof childNav !== 'boolean'),
  parentNav: PREGNANCY_MODULE_PARENT_NAV,
};

export const NBC_AND_PNC_NAVIGATION_MODULE: NavigationModule = {
  childNavs: [LOGFACE_PAGE_NAVIGATION],
  parentNav: NBC_AND_PNC_PARENT_NAV,
};

export const NUTRITION_MODULE: NavigationModule = {
  childNavs: [LOGFACE_PAGE_NAVIGATION],
  parentNav: NUTRITION_PARENT_NAVIGATION,
};

export const CLIENT_NAVIGATION_MODULE: NavigationModule = {
  childNavs: [
    ENABLE_CLIENT_PAGE && CLIENT_PAGE_NAVIGATION,
    ENABLE_HOUSEHOLD_PAGE && HOUSEHOLD_PAGE_NAVIGATION,
    ENABLE_ANC_PAGE && ANC_PAGE_NAVIGATION,
    ENABLE_CHILD_PAGE && CHILD_PAGE_NAVIGATION,
  ].filter((childNav): childNav is PageLink => typeof childNav !== 'boolean'),
  parentNav: CLIENT_MODULE_PARENT_NAV,
};

export const REPORT_NAVIGATION_MODULE: NavigationModule = {
  childNavs: [],
  parentNav: REPORT_MODULE_PARENT_NAV,
};
export const ADMIN_NAVIGATION_MODULE: NavigationModule = {
  childNavs: [
    ENABLE_USER_PAGE && USERS_PAGE_NAVIGATION,
    ENABLE_ROLE_PAGE && ROLES_PAGE_NAVIGATION,
    ENABLE_TEAM_PAGE && TEAMS_PAGE_NAVIGATION,
    ENABLE_LOCATION_PAGE && LOCATIONS_PAGE_NAVIGATION,
  ].filter<PageLink>(
    (childNav: PageLink | boolean): childNav is PageLink => typeof childNav !== 'boolean'
  ),
  parentNav: ADMIN_MODULE_PARENT_NAV,
};

/** Enable Clients Module from its child pages */
export const ENABLE_CLIENT_RECORDS_MODULE =
  ENABLE_ANC_PAGE || ENABLE_CLIENT_PAGE || ENABLE_HOUSEHOLD_PAGE || ENABLE_HOUSEHOLD_PAGE;
export type ENABLE_CLIENT_RECORDS_MODULE = typeof ENABLE_CLIENT_RECORDS_MODULE;

/** Enable Admin Module from its child pages */
export const ENABLE_ADMIN_MODULE =
  ENABLE_USER_PAGE || ENABLE_ROLE_PAGE || ENABLE_TEAM_PAGE || ENABLE_LOCATION_PAGE;
export type ENABLE_ADMIN_MODULE = typeof ENABLE_ADMIN_MODULE;
