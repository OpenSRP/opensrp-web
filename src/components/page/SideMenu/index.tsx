import * as React from 'react';
import { RouteComponentProps } from 'react-router';
import { Col, Row } from 'reactstrap';
import { ENABLE_REPORT_MODULE } from '../../../configs/env';
import {
  ENABLE_ANC_PAGE,
  ENABLE_CHILD_PAGE,
  ENABLE_CLIENT_PAGE,
  ENABLE_HOUSEHOLD_PAGE,
  ENABLE_LOCATION_PAGE,
  ENABLE_ROLE_PAGE,
  ENABLE_TEAM_PAGE,
  ENABLE_USER_PAGE,
} from '../../../configs/env';
import {
  ADMIN,
  ALL_CLIENTS,
  ANC,
  ANC_URL,
  CHILD,
  CHILD_URL,
  CLIENT_RECORDS,
  CLIENT_URL,
  HOUSEHOLD,
  HOUSEHOLD_URL,
  LOCATIONS,
  LOCATIONS_URL,
  REPORTS,
  REPORTS_URL,
  ROLE_URL,
  ROLES,
  TEAM_URL,
  TEAMS,
  USER_URL,
  USERS,
} from '../../../constants';
import SubMenu, { ModulePageLink, NavigationModule, PageLink, SubMenuProps } from './SubMenu';

import { IconProp } from '@fortawesome/fontawesome-svg-core';
import './index.css';

// Page links

export const CLIENT_PAGE_NAVIGATION: PageLink = {
  label: ALL_CLIENTS,
  url: CLIENT_URL,
};
export const HOUSEHOLD_PAGE_NAVIGATION: PageLink = {
  label: HOUSEHOLD,
  url: HOUSEHOLD_URL,
};
export const ANC_PAGE_NAVIGATION: PageLink = {
  label: ANC,
  url: ANC_URL,
};
export const CHILD_PAGE_NAVIGATION: PageLink = {
  label: CHILD,
  url: CHILD_URL,
};

export const USERS_PAGE_NAVIGATION: PageLink = {
  label: USERS,
  url: USER_URL,
};
export const ROLES_PAGE_NAVIGATION: PageLink = {
  label: ROLES,
  url: ROLE_URL,
};
export const TEAMS_PAGE_NAVIGATION: PageLink = {
  label: TEAMS,
  url: TEAM_URL,
};
export const LOCATIONS_PAGE_NAVIGATION: PageLink = {
  label: LOCATIONS,
  url: LOCATIONS_URL,
};

// icons
export const clientModuleNavIcon: IconProp = ['far', 'user'];
export const reportModuleNavIcon: IconProp = ['fas', 'chart-line'];
export const adminModuleNavIcon: IconProp = ['fas', 'cog'];

// Navigation links for navigation module. collapses when clicked to display the pageLinks

export const CLIENT_MODULE_PARENT_NAV: ModulePageLink = {
  icon: clientModuleNavIcon,
  label: CLIENT_RECORDS,
};
export const REPORT_MODULE_PARENT_NAV: ModulePageLink = {
  icon: reportModuleNavIcon,
  label: REPORTS,
  url: REPORTS_URL,
};
export const ADMIN_MODULE_PARENT_NAV: ModulePageLink = {
  icon: adminModuleNavIcon,
  label: ADMIN,
};

// nav module objects
export const CLIENT_NAVIGATION_MODULE: NavigationModule = {
  childNavs: [
    ENABLE_CLIENT_PAGE && CLIENT_PAGE_NAVIGATION,
    ENABLE_HOUSEHOLD_PAGE && HOUSEHOLD_PAGE_NAVIGATION,
    ENABLE_ANC_PAGE && ANC_PAGE_NAVIGATION,
    ENABLE_CHILD_PAGE && CHILD_PAGE_NAVIGATION,
  ].filter((childNav): childNav is PageLink => typeof childNav !== 'boolean'),
  parentNav: CLIENT_MODULE_PARENT_NAV,
};
export const REPORT_NAVIGATION_MODULE: NavigationModule = {
  childNavs: [],
  parentNav: REPORT_MODULE_PARENT_NAV,
};
export const ADMIN_NAVIGATION_MODULE: NavigationModule = {
  childNavs: [
    ENABLE_USER_PAGE && USERS_PAGE_NAVIGATION,
    ENABLE_ROLE_PAGE && ROLES_PAGE_NAVIGATION,
    ENABLE_TEAM_PAGE && TEAMS_PAGE_NAVIGATION,
    ENABLE_LOCATION_PAGE && LOCATIONS_PAGE_NAVIGATION,
  ].filter<PageLink>(
    (childNav: PageLink | boolean): childNav is PageLink => typeof childNav !== 'boolean'
  ),
  parentNav: ADMIN_MODULE_PARENT_NAV,
};

/** Enable Clients Module from its child pages */
export const ENABLE_CLIENT_RECORDS_MODULE =
  ENABLE_ANC_PAGE || ENABLE_CLIENT_PAGE || ENABLE_HOUSEHOLD_PAGE || ENABLE_HOUSEHOLD_PAGE;
export type ENABLE_CLIENT_RECORDS_MODULE = typeof ENABLE_CLIENT_RECORDS_MODULE;

/** Enable Admin Module from its child pages */
export const ENABLE_ADMIN_MODULE =
  ENABLE_USER_PAGE || ENABLE_ROLE_PAGE || ENABLE_TEAM_PAGE || ENABLE_LOCATION_PAGE;
export type ENABLE_ADMIN_MODULE = typeof ENABLE_ADMIN_MODULE;

/** interface for the local state for this component
 * @property {string} collapsedModuleLabel - label pointing to active navigation module
 */
export interface SideMenuState {
  collapsedModuleLabel: string;
}

const defaultSideMenuState: SideMenuState = {
  collapsedModuleLabel: '',
};

class SideMenu extends React.Component<{}, SideMenuState> {
  constructor(props: RouteComponentProps) {
    super(props);
    this.state = defaultSideMenuState;
  }

  public render() {
    const collapsedModuleLabel = this.state.collapsedModuleLabel;

    // props for subMenu components

    const clientSubMenuProps: SubMenuProps = {
      ...CLIENT_NAVIGATION_MODULE,
      collapsedModuleLabel,
      setCollapsedModuleLabel: this.setCollapsedModuleLabel,
    };
    const reportSubMenuProps: SubMenuProps = {
      ...REPORT_NAVIGATION_MODULE,
      collapsedModuleLabel,
      setCollapsedModuleLabel: this.setCollapsedModuleLabel,
    };
    const adminSubMenuProps: SubMenuProps = {
      ...ADMIN_NAVIGATION_MODULE,
      collapsedModuleLabel,
      setCollapsedModuleLabel: this.setCollapsedModuleLabel,
    };
    return (
      <div className="side-menu-container">
        <Row>
          <Col className="side-menu-extend">
            {ENABLE_CLIENT_RECORDS_MODULE && <SubMenu {...clientSubMenuProps} />}
            {ENABLE_REPORT_MODULE && <SubMenu {...reportSubMenuProps} />}
            {ENABLE_ADMIN_MODULE && <SubMenu {...adminSubMenuProps} />}
          </Col>
        </Row>
      </div>
    );
  }

  /** updates collapsedModuleLabel to label of the new navigation module that should be collapsed */
  private setCollapsedModuleLabel = (label: string) => {
    this.setState({
      collapsedModuleLabel: label,
    });
  };
}

export default SideMenu;
