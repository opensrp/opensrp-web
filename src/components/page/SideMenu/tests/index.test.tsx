import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { createBrowserHistory } from 'history';
import React from 'react';
import { Router } from 'react-router';
import snapshotDiff from 'snapshot-diff';
import SideMenu from '..';
import { CLIENT_RECORDS, PREGNANCY } from '../../../../constants';

const history = createBrowserHistory();

describe('components/page/SideMenu', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders without crashing', () => {
    shallow(
      <Router history={history}>
        <SideMenu />
      </Router>
    );
  });

  it('renders side menu correctly', () => {
    const wrapper = mount(
      <Router history={history}>
        <SideMenu />
      </Router>
    );
    /** client Collapse SubMenu renders correctly */
    expect(toJson(wrapper.find('Nav .side-collapse-nav').first())).toMatchSnapshot();
    wrapper.unmount();
  });

  it('manages state correctly', () => {
    const wrapper = mount(
      <Router history={history}>
        <SideMenu />
      </Router>
    );

    expect(wrapper.find('SideMenu').state('collapsedModuleLabel')).toEqual('');

    wrapper.find(`ul#${PREGNANCY.replace(' ', '-')}`).simulate('click');
    expect(wrapper.find('SideMenu').state('collapsedModuleLabel')).toEqual(PREGNANCY);
    wrapper.unmount();
  });

  it('sets the collapsedModuleLabel correctly from clicks on parentNavs', () => {
    // clicking changes the collapsedModuleLabel state and collapses
    // nav to reveal child navigation

    const wrapper = mount(
      <Router history={history}>
        <SideMenu />)
      </Router>
    );

    // starts with sub-menu as closed, not collapsed
    const clientChildNav = wrapper.find('div.collapse.show a[href="/clients"]');
    expect(clientChildNav.length).toEqual(0);
    const beforeClickWrapper = toJson(wrapper);

    // clicking on a parent nav changes the collapsedState for that navigation module
    const PregnancyNav = wrapper.find(`ul#${PREGNANCY.replace(' ', '-')}`);
    expect(PregnancyNav.length).toEqual(1);
    PregnancyNav.simulate('click');
    wrapper.update();
    const afterClickWrapper = toJson(wrapper);

    // isOpen value for collapsible div holding child navs changes from false to true
    expect(snapshotDiff(beforeClickWrapper, afterClickWrapper)).toMatchSnapshot('Everything');

    wrapper.unmount();
  });
});
