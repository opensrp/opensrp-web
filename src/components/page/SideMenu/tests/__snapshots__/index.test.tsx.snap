// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/page/SideMenu renders side menu correctly 1`] = `
<ul
  className="side-collapse-nav nav"
  id="Home"
  onClick={[Function]}
>
  <NavItem
    tag="li"
  >
    <li
      className="nav-item"
    >
      <NavLink
        className="nav nav-link side-nav-item active-collapse-menu"
        to="/"
      >
        <Link
          aria-current="page"
          className="nav nav-link side-nav-item active-collapse-menu active"
          style={Object {}}
          to="/"
        >
          <a
            aria-current="page"
            className="nav nav-link side-nav-item active-collapse-menu active"
            href="/"
            onClick={[Function]}
            style={Object {}}
          >
            <FontAwesomeIcon
              border={false}
              className=""
              fixedWidth={false}
              flip={null}
              icon={
                Array [
                  "fas",
                  "home",
                ]
              }
              inverse={false}
              listItem={false}
              mask={null}
              pull={null}
              pulse={false}
              rotation={null}
              size="lg"
              spin={false}
              symbol={false}
              title=""
              transform={null}
            />
            <span
              className="collapse-menu-title"
            >
               
              Home
               
            </span>
          </a>
        </Link>
      </NavLink>
    </li>
  </NavItem>
</ul>
`;

exports[`components/page/SideMenu sets the collapsedModuleLabel correctly from clicks on parentNavs: Everything 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -122,11 +122,11 @@
                  <div
                    className=\\"side-menu-extend col\\"
                  >
                    <withRouter(SubMenu)
                      childNavs={Array []}
-                     collapsedModuleLabel=\\"\\"
+                     collapsedModuleLabel=\\"Pregnancy\\"
                      key=\\"0\\"
                      parentNav={
                        Object {
                          \\"icon\\": Array [
                            \\"fas\\",
@@ -138,11 +138,11 @@
                      }
                      setCollapsedModuleLabel={[Function anonymous]}
                    >
                      <SubMenu
                        childNavs={Array []}
-                       collapsedModuleLabel=\\"\\"
+                       collapsedModuleLabel=\\"Pregnancy\\"
                        history={
                          Object {
                            \\"action\\": \\"POP\\",
                            \\"block\\": [Function block],
                            \\"createHref\\": [Function createHref],
@@ -320,11 +320,11 @@
                            \\"label\\": \\"Analysis\\",
                            \\"url\\": \\"/analysis\\",
                          },
                        ]
                      }
-                     collapsedModuleLabel=\\"\\"
+                     collapsedModuleLabel=\\"Pregnancy\\"
                      key=\\"1\\"
                      parentNav={
                        Object {
                          \\"icon\\": Array [
                            \\"far\\",
@@ -351,11 +351,11 @@
                              \\"label\\": \\"Analysis\\",
                              \\"url\\": \\"/analysis\\",
                            },
                          ]
                        }
-                       collapsedModuleLabel=\\"\\"
+                       collapsedModuleLabel=\\"Pregnancy\\"
                        history={
                          Object {
                            \\"action\\": \\"POP\\",
                            \\"block\\": [Function block],
                            \\"createHref\\": [Function createHref],
@@ -496,11 +496,11 @@
                          <Collapse
                            appear={false}
                            enter={true}
                            exit={true}
                            in={false}
-                           isOpen={false}
+                           isOpen={true}
                            mountOnEnter={false}
                            onEnter={[Function noop]}
                            onEntered={[Function noop]}
                            onEntering={[Function noop]}
                            onExit={[Function noop]}
@@ -512,11 +512,11 @@
                          >
                            <Transition
                              appear={false}
                              enter={true}
                              exit={true}
-                             in={false}
+                             in={true}
                              mountOnEnter={false}
                              onEnter={[Function noop]}
                              onEntered={[Function bound onEntered]}
                              onEntering={[Function bound onEntering]}
                              onExit={[Function bound onExit]}
@@ -524,12 +524,16 @@
                              onExiting={[Function bound onExiting]}
                              timeout={350}
                              unmountOnExit={false}
                            >
                              <div
-                               className=\\"collapse\\"
-                               style={Object {}}
+                               className=\\"collapsing\\"
+                               style={
+                                 Object {
+                                   \\"height\\": 0,
+                                 }
+                               }
                              >
                                <Nav
                                  key=\\"child-nav-0\\"
                                  tag=\\"ul\\"
                                  vertical={false}"
`;
